File: tkpdshop/src/main/java/com/tokopedia/shop/note/view/adapter/ShopNoteAdapterTypeFactory.java
package com.tokopedia.shop.note.view.adapter;

import android.view.View;

import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory;
import com.tokopedia.abstraction.base.view.adapter.model.EmptyModel;
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder;
import com.tokopedia.abstraction.base.view.adapter.viewholders.EmptyViewHolder;
import com.tokopedia.shop.note.view.adapter.viewholder.ShopNoteViewHolder;
import com.tokopedia.shop.note.view.model.ShopNoteViewModel;

/**
 * Created by alvarisi on 12/7/17.
 */

public class ShopNoteAdapterTypeFactory extends BaseAdapterTypeFactory {

    private final EmptyViewHolder.Callback emptyNoteOnClickListener;

    public ShopNoteAdapterTypeFactory(EmptyViewHolder.Callback emptyNoteOnClickListener) {
        this.emptyNoteOnClickListener = emptyNoteOnClickListener;
    }

    @Override
    public int type(EmptyModel viewModel) {
        return EmptyViewHolder.LAYOUT;
    }

    public int type(ShopNoteViewModel shopNoteViewModel) {
        return ShopNoteViewHolder.LAYOUT;
    }

    @Override
    public AbstractViewHolder createViewHolder(View view, int viewType) {
        if (viewType == EmptyViewHolder.LAYOUT) {
            return new EmptyViewHolder(view, emptyNoteOnClickListener);
        }  else if (viewType == ShopNoteViewHolder.LAYOUT) {
            return new ShopNoteViewHolder(view);
        } else {
            return super.createViewHolder(view, viewType);
        }
    }
}

File: features/merchant/shop_page/src/main/java/com/tokopedia/shop/note/view/adapter/ShopNoteAdapterTypeFactory.kt
package com.tokopedia.shop.note.view.adapter

import android.view.View
import com.tokopedia.abstraction.base.view.adapter.Visitable
import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory
import com.tokopedia.abstraction.base.view.adapter.model.EmptyModel
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder
import com.tokopedia.shop.info.view.adapter.viewholder.ShopNoteEmptyViewHolder
import com.tokopedia.shop.note.view.adapter.viewholder.ShopNoteViewHolder
import com.tokopedia.shop.note.view.model.ShopNoteUiModel

class ShopNoteAdapterTypeFactory(val onNoteClicked: ShopNoteViewHolder.OnNoteClicked): BaseAdapterTypeFactory(){

    fun type(shopNoteUiModel: ShopNoteUiModel?) = ShopNoteViewHolder.LAYOUT

    override fun type(viewModel: EmptyModel) = ShopNoteEmptyViewHolder.LAYOUT

    override fun createViewHolder(parent: View, type: Int): AbstractViewHolder<out Visitable<*>> {
        return when(type){
            ShopNoteEmptyViewHolder.LAYOUT -> ShopNoteEmptyViewHolder(parent)
            ShopNoteViewHolder.LAYOUT -> ShopNoteViewHolder(parent, onNoteClicked)
            else -> super.createViewHolder(parent, type)
        }
    }
}